core/types/accounts/account.go

```diff
@@ -521,6 +556,188 @@ func (a *Account) decode(enc []byte, forStorage bool) error {
 	return nil
 }
 
+func (a *Account) DecodeForStorage(enc []byte) error {
+	pos := 1
+	a.Initialised = true
+	a.Nonce = 0
+	a.Balance.SetInt64(0)
+	a.Root = emptyRoot
+	a.CodeHash = emptyCodeHash
+	a.StorageSize = 0
+	a.HasStorageSize = false
+
+	var ch codec.CborHandle
+	var buffer bytes.Buffer
+	decoder := codec.NewDecoder(&buffer, &ch)
+	if len(enc) == 0 {
+		return nil
+	}
+
+	var fieldSet = enc[0]
+	if fieldSet&1 > 0 {
+		nonceBytes, newPos := decodeLengthForStorage(enc, pos)
```

Comment:
You can remove `decodeLengthForStorage` completely, and simply use `decoder` to decode things